In [1]:
import pandas as pd

In [2]:
## reading data
data1 = pd.read_excel(r"G:\NeenOpal\Records Matching Task.xlsx", sheet_name = "data1")

data2 = pd.read_excel(r"G:\NeenOpal\Records Matching Task.xlsx", sheet_name = "data2")

In [3]:
data1.shape, data2.shape

((9478, 3), (9479, 3))

In [5]:
data1[:2], data2[:2]

(         Order ID       Product ID  Qty
 0  CA-2014-100006  TEC-PH-10002075    3
 1  CA-2014-100090  FUR-TA-10003715    3,
          Order ID       Product ID  Qty
 0  CA-2014-100090  FUR-TA-10003715    3
 1  CA-2014-100090  OFF-BI-10001597    6)

In [6]:
## creating key for the both datasets (combining Order ID & Product ID)

data1["Key"] = data1["Order ID"].astype(str) + "_" + data1["Product ID"]

data2["Key"] = data2["Order ID"].astype(str) + "_" + data2["Product ID"]

In [7]:
data1[:2], data2[:2]

(         Order ID       Product ID  Qty                             Key
 0  CA-2014-100006  TEC-PH-10002075    3  CA-2014-100006_TEC-PH-10002075
 1  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090_FUR-TA-10003715,
          Order ID       Product ID  Qty                             Key
 0  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090_FUR-TA-10003715
 1  CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090_OFF-BI-10001597)

## Question 1: How to identify the Records (Order ID + Product ID combination) present in data1 but missing in data2 (Specify the number of records missing in your answer)

#### Answer is 507 

### Approach: 
- Left merge data2 with data1 (i.e. data1 "left join" data2).
- Using _indicator_ parameter to find out which rows are present only in data1 (left_only) and which are present in both the datasets (both).
- Selecting only rows that have "left_only".

#### Note:
- Merging on all the columns to avoid creation of multiple columns with same value. 

In [14]:
data1_left_join = data1.merge(data2, on = ["Order ID", "Product ID", "Qty","Key"], how = "left", indicator = True)

In [15]:
data1_left_join[:2]

Order ID       Product ID  Qty                             Key  \
0  CA-2014-100006  TEC-PH-10002075    3  CA-2014-100006_TEC-PH-10002075   
1  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090_FUR-TA-10003715   

      _merge  
0  left_only  
1       both

In [16]:
data1_only = data1_left_join[data1_left_join["_merge"] == "left_only"]

In [17]:
data1_only.shape

(507, 5)

## Question 2: How to identify the Records (Order ID + Product ID combination) missing in data1 but present in data2 (Specify the number of records missing in your answer)

#### Answer is 508

### Approach (similar to Qn 1): 
- Left merge data1 with data2  (i.e. data2 "left join" data1).
- Using _indicator_ parameter to find out which rows are present only in data1 (left_only) and which are present in both the datasets (both).
- Selecting only rows that have "left_only".

#### Note:
- Merging on all the columns to avoid creation of multiple columns with same value. 

In [20]:
data2_left_join = data2.merge(data1, on = ["Order ID", "Product ID", "Qty","Key"], how = "left", indicator = True)

data2_left_join[:2]

Order ID       Product ID  Qty                             Key _merge
0  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090_FUR-TA-10003715   both
1  CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090_OFF-BI-10001597   both

In [21]:
data2_only = data2_left_join[data2_left_join["_merge"] == "left_only"]

data2_only.shape

(508, 5)

## Question 3: Find the Sum of the total Qty of Records missing in data1 but present in data2

#### Answer is 1956

#### Approach: 
- The final dataset orginated in Question 2 (i.e. data2_only) is our data of interest.
- Applying pandas sum() to find the sum of the column values.

In [22]:
data2_only.head()

Order ID       Product ID  Qty                             Key  \
11  CA-2014-100706  TEC-AC-10001314    2  CA-2014-100706_TEC-AC-10001314   
14  CA-2014-100762  OFF-PA-10001815    3  CA-2014-100762_OFF-PA-10001815   
30  CA-2014-101427  OFF-AR-10002257    3  CA-2014-101427_OFF-AR-10002257   
56  CA-2014-102652  FUR-FU-10001918    7  CA-2014-102652_FUR-FU-10001918   
63  CA-2014-102869  OFF-PA-10000788    3  CA-2014-102869_OFF-PA-10000788   

       _merge  
11  left_only  
14  left_only  
30  left_only  
56  left_only  
63  left_only

In [25]:
## checking if there is any missing value in the column "Qty"
data2_only["Qty"].isna().sum()

0

In [27]:
data2_only["Qty"].sum()

1956

## Question 4: Find the total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2


#### Answer is 8971

#### Approach:
- Applying inner join on data1 and data2 to get the common rows.
- Using the columns along with "Key" as primary key to combine them.
- Finding the shape to get the number of rows, hence finding the number of unique records.

#### Alternative approach:
- We have size of data1 (9478).
- We also have the size of data that is present in data1 and not present in data2 (507). 
- So the number of rows present in both data1 and data2 = _9478 - 507 = 8971_

In [38]:
combined_data = data1.merge(data2, on = ["Order ID", "Product ID", "Qty","Key"], how = "inner")

In [39]:
combined_data.shape

(8971, 4)